pkg/trace/otel.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		log.Error(ctx, "Unable to start trace exporter", events.NewErrorInfo(err))
+	}
+
+	r, err := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String(serviceName),
+		),
+	)
+	if err != nil {
+		log.Error(ctx, "Unable to configure trace provider", events.NewErrorInfo(err))
+	}
+
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithBatcher(exp),
+		sdktrace.WithResource(r),
+		// otel accepts sample rates as a fractions of traces >= 1 will always sample < 0 will never sample
+		sdktrace.WithSampler(sdktrace.TraceIDRatioBased(t.Otel.SamplePercent/100)),
+		sdktrace.WithSpanProcessor(Annotator{globalTags: t.GlobalTags}),
+	)
+
+	otel.SetTracerProvider(tp)
+
+	otel.SetTextMapPropagator(
+		propagation.NewCompositeTextMapPropagator(
+			propagation.TraceContext{},
+			propagation.Baggage{},
+		),
+	)
+
+	t.serviceName = serviceName
+	t.tracerProvider = tp
+
+	t.tracerProvider.Tracer(serviceName)
+
+	return nil
+}
+
+// Deprecated: Use closeTracer() instead.
+func (t *otelTracer) endTracing() {
+	t.closeTracer(context.TODO())
+}
+
+func (t *otelTracer) closeTracer(ctx context.Context) {
+	if t.tracerProvider == nil {
+		return
+	}
+
+	t.tracerProvider.ForceFlush(ctx)
+	err := t.tracerProvider.Shutdown(ctx)
+	if err != nil {
+		log.Error(ctx, "Unable to stop otel tracer", events.NewErrorInfo(err))
+	}
+}
+
+func (t *otelTracer) startTrace(ctx context.Context, name string) context.Context {
+	tracer := otel.GetTracerProvider().Tracer(t.serviceName)
+	ctx, _ = tracer.Start(ctx, name)
+
+	return ctx
+}
+
+func (t *otelTracer) id(ctx context.Context) string {
+	if span := trace.SpanFromContext(ctx); span != nil && span.SpanContext().TraceID().IsValid() {
+		return "oteltrace_" + span.SpanContext().TraceID().String()
+	}
+	return ""
+}
+
+func (t *otelTracer) startSpan(ctx context.Context, name string) context.Context {
+	tracer := otel.GetTracerProvider().Tracer(t.serviceName)
+	ctx, _ = tracer.Start(ctx, name)
+
+	return ctx
+}
+
+func (t *otelTracer) startSpanAsync(ctx context.Context, name string) context.Context {
+	tracer := otel.GetTracerProvider().Tracer(t.serviceName)
+	ctx, _ = tracer.Start(ctx, name)
+
+	return ctx
+}
+
+func (t *otelTracer) end(ctx context.Context) {
+	if span := trace.SpanFromContext(ctx); span != nil {
+		span.End()
+	}
+}
+
+func (t *otelTracer) addInfo(ctx context.Context, args ...log.Marshaler) {
+	if span := trace.SpanFromContext(ctx); span != nil {
+		for _, f := range args {
+			logf.Marshal("", f, func(key string, value interface{}) {
+				span.SetAttributes(attribute.String(key, fmt.Sprintf("%v", value)))
+			})
+		}
+	}
+}
+
+func (t *otelTracer) spanID(ctx context.Context) string {
+	if span := trace.SpanFromContext(ctx); span != nil && span.SpanContext().SpanID().IsValid() {
+		return span.SpanContext().SpanID().String()
+	}
+	return ""
+}
+
+func (t *otelTracer) parentID(ctx context.Context) string {
+	return ""
```

Comment:
This doesn't look right.  If this is intentional, maybe add a comment?

Reply:
Updated with deprecation comments.